Documentation/Yarn-Spec.md

```diff
@@ -0,0 +1,636 @@
+Parentheses must be treated as if they are a single operand in other operations.
+Parentheses start with the open bracket symbol `(` and can have any expression inside of them before being closed with the closing bracket symbol `)`.
+`2 * (3 + $coins)` is an example of an expression with a parentheses operation, in this case the `3 + $coins` component must be resolved into a value before being able to be multiplied by two.
+
+The `+` operator when operating on strings is not addition in the mathematical sense but concatenation.
+
+#### Supported types in operations
+
+The following table shows the compatible types for each binary operation and must be supported:
+
+|          | + | - | * | / | % | == | != | gt | lt | gte | lte | \|\| | && |
+|----------|---|---|---|---|---|----|----|----|----|-----|-----|------|----|
+| numbers  | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ | ✅  | ✅ | ✅ | ✅ |
+| strings  | ✅ | ❌ | ❌ | ❌ | ❌ | ✅ | ✅ | ❌ | ❌ | ❌  | ❌ | ❌ | ❌ |
+| booleans | ❌ | ❌ | ❌ | ❌ | ❌ | ✅ | ✅ | ❌ | ❌ | ❌  | ❌ | ✅ | ✅ |
+
+The following table shows the compatible types for each unary operation and must be supported:
+
+|          | ! | - |
+|----------|---|---|
+| numbers  | ❌ | ✅ |
+| strings  | ❌ | ❌ |
+| booleans | ✅ | ❌ |
+
+Operations between different or incompatible types is unspecified but should not be supported due to the potential confusion.
+If however they are supported existing behaviour around transitivity, commutativity, and symmetry of operations should be respected.
+For example if `"hello" < 5` is `true` then `5 > "hello"` should also be `true`.
+Likewise `("hello" + 2) < 5` resulted in `true` then `(2 + "hello") < 5` should also be `true`.
+This taken to the extreme should mean that `1 + "hello" == "hello + 1` should evaluate to `true` which is confusing to most people, hence the recommendation against supporting operations between disparate types.
+
+#### Operation Output
+
+The following table shows the expected output type of each operation based on its operand type:
+
+|         | +      | -      | *      | /      | %      | ==      | !=      | gt      | lt      | gte     | lte     | \|\|    | &&      | !       | unary minus |
+|---------|--------|--------|--------|--------|--------|---------|---------|---------|---------|---------|---------|---------|---------|---------|-------------|
+| number  | number | number | number | number | number | boolean | boolean | boolean | boolean | boolean | boolean | boolean | boolean |         | number      |
+| string  | string |        |        |        |        | boolean | boolean |         |         |         |         |         |         |         |             |
+| boolean |        |        |        |        |        | boolean | boolean |         |         |         |         | boolean | boolean | boolean |             |
+
+#### Order of Operations
+
+The order of operations is as follows:
+
+1. `()`
+1. `!`, `unary minus`
```

Comment:
Weird to spell out one and not the other. Say "`!`, `-` (unary)" or spell out each category in each numbered item in this list, e.g. "Unary operators: `!`,` -`", "Comparison operators: `==`...", etc.